dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java

```diff
@@ -1,27 +1,39 @@
 package datadog.trace.instrumentation.kotlin.coroutines;
 
+import static datadog.trace.instrumentation.kotlin.coroutines.CoroutineContextHelper.getJob;
+
 import datadog.trace.bootstrap.instrumentation.api.AgentScope;
 import datadog.trace.bootstrap.instrumentation.api.AgentTracer;
 import datadog.trace.bootstrap.instrumentation.api.ScopeState;
+import kotlin.Unit;
 import kotlin.coroutines.CoroutineContext;
+import kotlin.jvm.functions.Function1;
 import kotlin.jvm.functions.Function2;
 import kotlinx.coroutines.ThreadContextElement;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-public class ScopeStateCoroutineContext implements ThreadContextElement<ScopeState> {
+public class ScopeStateCoroutineContext
+    implements ThreadContextElement<ScopeState>, Function1<Throwable, Unit> {
 
-  private static final Key<ScopeStateCoroutineContext> KEY = new ContextElementKey();
+  public static final Key<ScopeStateCoroutineContext> KEY = new ContextElementKey();
   private final ScopeState coroutineScopeState;
-  @Nullable private ContinuationHandler continuationHandler;
+  @Nullable private AgentScope.Continuation continuation;
+  @Nullable private AgentScope continuationScope;
```

Comment:
I think that these variables can still be accessed concurrently by multiple threads and should be using `AtomicReference`

Reply:
Yeah, you're right, I'll change that

Reply:
@bantonsson 
Altho, now that I went through the code again and through thread context element docs (mostly the comments [here](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/jvm/src/ThreadContextElement.kt)), I really don't see how there could be any concurrent writes there.

First call to `updateThreadContext` will always be "synchronous" it can't happen in parallel with `restoreThreadContext`. We initialize those fields on the first call only, later on there are only reads.

Then there's also `coroutineScopeState`, but we don't access it at all on call to `restoreThreadContext`, instead we write the `ScopeStack` to the thread local variable using `ScopeState`. 

The only possible reason why we can have concurrent calls to `updateThreadContext`/`restoreThreadContext` is if there's a bug somewhere else. And there was one! :slightly_smiling_face: 

When you call `withTimeout` internally it creates a new coroutine of type `TimoutCoroutine` and it turns out it's handled a bit differently than others. First of all when it is created, there's no invocation of `CoroutineContextKt.newCoroutineContext` (so we weren't creating an instance of `ScopeStateCoroutineContext` for it and it was inheriting this item from the parent context, probably this is what was causing concurrent access to the context element). Second problem was that it is handled a bit differently from others, in a way that it's not guaranteed to start, in some cases it just runs the code block without dispatch. In such cases the on completion callback won't run. Luckily, there's another method available in `JobSupport` class that we can instrument to guarantee on completion callback execution even when `TimeoutCoroutine` hasn't been dispatched - `onCompletionInternal`. 

So here 27e144ecb31d9dc1a14ca55483a49f87f3b657b8 I did a few changes:

- `ScopeStateCoroutineContext` is now crated on invocation of `AbstractCoroutine` constructor instead of `CoroutineContextKt.newCoroutineContext`, this way we guarantee every coroutine started will have a new instance of `ScopeStateCoroutineContext` and will not inherit it from the parent coroutine, so we shouldn't have any concurrent access to `ScopeStateCoroutineContext` instance.
- `maybeCloseScopeAndCancelContinuation` is now guaranteed to be invoked when coroutine transitions to a terminal state (`Cancelled` or `Completed`)
- also removed the optional scope propagation I added here (48922a70edf550e577cd4850c2dc1108d06449c5) bofre I saw [your comment](https://github.com/monosoul/dd-trace-java/pull/3#discussion_r1087535929) :slightly_smiling_face:

Reply:
Nice catch with the `TimeoutCoroutine` and `newCoroutineContext`. That explains why multiple threads were accessing the same instance.